src/Native/Unix/System.Native/pal_random.cpp

```diff
@@ -20,72 +20,51 @@ extern "C" void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* b
 {
     assert(buffer != NULL);
 
-    int rand_des;
-    char buf;
+    static volatile int rand_des = -1;
     long num = 0;
-
-    static bool sMissingDevRandom;
     static bool sMissingDevURandom;
     static bool sInitializedMRand;
 
-    int i = 0;
-
-    if (i < bufferLength && !sMissingDevRandom)
+    if (!sMissingDevURandom)
     {
-        // request non-blocking access to avoid hangs if the /dev/random is exhausted
-        // or just simply broken
-        if ((rand_des = open("/dev/random", O_RDONLY | O_NONBLOCK)) == -1)
+        if (rand_des == -1)
         {
-            if (errno == ENOENT)
+            int fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
```

Comment:
Is it thread safe to read concurrently from this now shared fd?

Reply:
See the "Bugs" section in http://man7.org/linux/man-pages/man2/read.2.html. It says that from Linux >= 3.14, it is thread safe. 
On Linux < 3.14, the doc says that multiple threads can possibly obtain blocks that overlap. 

But looking at the kernel sources, the random data extraction has a spinlock around it, so it should be safe even on the older Linux kernels.

Reply:
What about on macOS / BSD?

Reply:
As for the read functions, according to their man pages, they should both conform to POSIX 1 standard, like on Linux. That means that the operation should thread safe there too. However, POSIX talks only about regular files, not pseudo files like /dev/urandom.

Looking at FreeBSD sources, there is a lock around reading from the random generator in the /dev/urandom handler.

And the same is true for macOS source code (checked in the 10.12 version).

Reply:
Ok. Thanks.